### ISSUE TITLE
framing.character_delimited.delimiter no longer works

### ISSUE THREAD

**Issue by filterfish** [vectordotdev/vector#10814]
#### Working:
```
vector 0.18.1 (x86_64-unknown-linux-gnu c4adb60 2021-11-30)
```

#### Not working:
```
vector 0.19.0 (x86_64-unknown-linux-gnu da60b55 2021-12-28)
```

### Vector Configuration File

```toml
[sources.source]
type = "stdin"
framing.method = "character_delimited"
framing.character_delimited.delimiter = "\u0000"

[sinks.sink]
type = "console"
inputs = ["source"]
encoding.codec = "json"
```

#### Working version (vector 0.18.1)

When using version `vector 0.18.1 (x86_64-unknown-linux-gnu c4adb60 2021-11-30)` vector ouputs what I expect:

```
$ printf "hey\0" | vector -c vector.stdin.toml
2022-01-12T07:18:02.500104Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info"
2022-01-12T07:18:02.500496Z  INFO vector::app: Loading configs. paths=["vector.stdin.toml"]
2022-01-12T07:18:02.521159Z  INFO vector::topology::running: Running healthchecks.
2022-01-12T07:18:02.521491Z  INFO vector::sources::stdin: Capturing STDIN.
2022-01-12T07:18:02.521543Z  INFO vector::topology::running: Starting source. key=source
2022-01-12T07:18:02.521591Z  INFO vector::topology::builder: Healthcheck: Passed.
2022-01-12T07:18:02.522119Z  INFO vector::topology::running: Starting sink. key=sink
2022-01-12T07:18:02.522353Z  INFO vector: Vector has started. debug="false" version="0.18.1" arch="x86_64" build_id="c4adb60 2021-11-30"
2022-01-12T07:18:02.522456Z  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
2022-01-12T07:18:02.522604Z  INFO vector::shutdown: All sources have finished.
2022-01-12T07:18:02.522598Z  INFO source{component_kind="source" component_id=source component_type=stdin component_name=source}: vector::sources::stdin: Finished sending.
2022-01-12T07:18:02.522637Z  INFO vector: Vector has stopped.
{"host":"pike","message":"hey","source_type":"stdin","timestamp":"2022-01-12T07:18:02.522411555Z"}
```

#### Non-working version (vector 0.19.0)
When using version `vector 0.19.0 (x86_64-unknown-linux-gnu da60b55 2021-12-28)` I get an configuration error:

```
2022-01-12T07:21:25.761201Z  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info,buffers=info"
2022-01-12T07:21:25.761416Z  INFO vector::app: Loading configs. paths=["vector.stdin.toml"]
2022-01-12T07:21:25.762699Z ERROR vector::cli: Configuration error. error=invalid type: string "\u{0}", expected u8 for key `sources.source` at line 7 column 1
```
### Additional Context

I've tried every thing I can thing of when setting `framing.character_delimited.delimiter` but it always results in the same error. Including (this is not a comprehensive list):

* `framing.character_delimited.delimiter = "\0"`
* `framing.character_delimited.delimiter = "\\0"`
* `framing.character_delimited.delimiter = '\0'`

* `framing.character_delimited.delimiter = "\u0"`

I was basically cargo culting at this stage!

**Comment by spencergilbert** [comment]
cc @pablosichert

**Comment by pablosichert** [comment]
Thanks for the report @filterfish!

The `character_delimited` framing internally uses a single byte character only - previously we allowed any `char` (which in Rust is 4 byte wide) to be provided to the configuration, effectively truncating all but the last byte. This can lead to unexpected results, so we decided to only allow one byte characters in the configuration moving forward.

In https://github.com/vectordotdev/vector/pull/10570 we fixed this behavior but accidentally introduced a bug by not updating the config deserializing accordingly. Will keep you updated for a fix!

**Comment by jszwedko** [comment]
@pablosichert do you think it'd be possible to permit the null byte as a delimiter? It is an ASCII character, but I'm not 100% sure how you'd specify it in TOML.

**Comment by filterfish** [comment]
> In #10570 we fixed this behavior but accidentally introduced a bug by not updating the config deserializing accordingly.

I saw it mentioned in the release notes but figured a regression had crept in somewhere!

> Will keep you updated for a fix!

Good one! I'm ok using 0.18 for the time being and it'll be a while before this goes into production so no rush on my account.

**Comment by filterfish** [comment]
> @pablosichert do you think it'd be possible to permit the null byte as a delimiter? It is an ASCII character, but I'm not 100% sure how you'd specify it in TOML.

I tried that but it didn't work in the either 0.18 or 0.19.

I used the following command to generate the null (I'm sure there are better ways but this is the one that sprung to mind first):

```bash
printf "framing.character_delimited.delimiter = '\0'\n" >> vector.toml
```

and then moved it to the right place in vi.

**Comment by pablosichert** [comment]
The fix will go out in `0.19.1` and you'll be able to set

```
framing.character_delimited.delimiter = "\u0000"
```

again.

> do you think it'd be possible to permit the null byte as a delimiter? It is an ASCII character, but I'm not 100% sure how you'd specify it in TOML.

@jszwedko you'll be able to config ASCII characters via string as before. E.g. `"\n"`, `"x"` or `"\u0000"`. The config will fail to parse (opposed to the previous behavior where it truncated the upper bytes) and throw an error if you provide characters outside of the ASCII range, e.g. `"ä"` or if they are not a character, e.g. `0`.

> This works, in 0.19 anyway:
> ```
> framing.character_delimited.delimiter = 0x0
> ```

While that is true, I'd caution you against using that as this will rely on undocumented behavior and will not be supported anymore 😄
